### DISCUSSION TITLE
用weblate翻译时commit太多怎么办?

### DISCUSSION THREAD

**Discussion by exzhawk** [CFPAOrg/Minecraft-Mod-Language-Package#1018] [General]
### 前情提要
weblate 被魔改用于在zen模式处理建议 https://github.com/exzhawk/weblate/commit/ce5106970502e43825340ece0608beaa90b74c84
在zen模式比较容易连续进行 "接受并修改" 多个条目

### 出现的现象
迅速出现多个commit
![image](https://user-images.githubusercontent.com/7038938/110970161-6c482d00-8394-11eb-9ba8-064c73cf5dc2.png)

### 出现的原因
"接受并修改" 操作分为两步: 
1. 用户A接受了用户B的建议, 目前翻译结果字符串为用户B提交的内容
2. 用户A对目前的翻译结果进行了修改

因为 weblate 为了尽量保持作者信息, 需要体现出, 是用户B的建议先被接受, 然后用户A才进行了修改. 为此, 必须有两个commit才能体现出来, 如此连续进行此操作就会导致出现多个commit

### 导致的后果

1. GitHub 上commit过多, 看起来不方便
2. git repo 里 commit 过多, 进行一些操作的时候会拖慢速度
3. 由于进行commit操作的时候, 用户同时也在操作zen界面, 有时候会导致抢占weblate文件锁失败, 可能导致部分zen操作丢失(未验证).

### 可能的解决方法
1. 啥也不干 遇到困难睡大觉. 反正不是致命问题.
2. 放弃保持完整的credit信息, 在zen界面里, "接受并修改" 时不进行"接受"操作, 直接(自动)复制内容到翻译框并进行修改(并自动删除建议). 会在"接受并修改"时丢失被"接受"的那个作者信息. 最后只会体现出是点了"接受并修改"的用户贡献了此行翻译.
3. 关闭push on commit功能, 在push之前, 把两人的操作的commit分离开, 分别squash之后push. 暂未实现. 不能解决后果3. 可能加大weblate到github的同步延迟.

**Comment by WuzgXY-GitHub** [comment]
讨论后的后果 1 解决方法：
(1) 对于代码 commit，可以打开一个新的 branch，今后只存储代码 commit，需要时 push 回主 branch
(2) 对于翻译 commit，使用 blame 功能

可能的解决方法 4：
另开一库，Weblate commit 统一被收至该库，周期性 push 回本库并同步本库的修改；push 时仅保留 1 条 commit，并在 description 中写明对应的 Weblate commit（我不是很了解 parent 机制，但我觉得这么做不会出问题）

**Comment by dovisutu** [comment]
我的想法是，尽量要保持作者信息（毕竟有个contributors界面，虽然现在分的1.16分支已经把它快弄坏掉了），但是过于详细的“原来提交的是什么，怎么被改的”或许只用在wl上记录，commit时使用co-authored-by标记，这样子至少可以把commit减少一点；
如果要进一步减少或许可以记录每5个（例如）suggestion单元合并为一个大的单元，每次译者执行到末尾把积压的单元再次commit

**Reply by WuzgXY-GitHub** [reply]
Co-authored by 似乎会产生错误的行数统计

**Comment by julliazmarketing-lgtm** [comment]
Why It Happens

Weblate uses lazy commits, meaning it groups multiple translation edits by the same author into a single commit—but only when certain conditions are met (e.g., another author edits the string, a merge happens, an explicit commit is requested, or a certain time threshold is passed) 
Weblate
.

With many components or frequent edits, this can still lead to a flood of commits.

How to Reduce Commit Spam – Quick Fixes

Enable the "Squash Git commits" Add-on [https://minecrftgeek.com](https://minecrftgeek.com/
)
The Squash Git commits add-on can merge multiple Weblate edits into a single commit (e.g., per language or all changes together) 
Weblate
+1
.

If it’s not working, it may be due to bugs in older Weblate versions. Upgrading resolved issues for others—for instance, moving from version 3.5.1 to 3.7 made the add-on work correctly 
Stack Overflow
.

Adjust Lazy Commit Timing

In the component configuration, set "Age of changes to commit" to a higher value so Weblate waits longer before committing, reducing commit frequency 
Weblate
.

You can also explicitly schedule periodic commits using Django’s periodic task system—for example, every 2 minutes—via weblate.trans.tasks.commit_pending 
Weblate
.

Manual Triggering of Pending Commits

Use management commands to batch commits selectively:

weblate commitgit <component> — flushes all pending changes in the component to Git.

weblate commit_pending --age HOURS — commits changes older than the specified hours 
Weblate
.

Summary Table
Strategy	What It Does
Squash Git commits add-on	Merges multiple edits into fewer, cleaner commits
Increase commit age threshold	Delays commits to aggregate more changes
Use commit_pending command	Manually trigger batch commits
Upgrade Weblate	Fixes known add-on bugs (e.g., squashing failures)
Final Take

To avoid overwhelming your Git history:

Install and enable the Squash Git commits add-on (make sure your Weblate version supports it).

Modify the lazy commit "age" setting to delay committing.

Use management commands for manual control when needed.

Need help enabling the add-on or configuring periodic tasks? Happy to walk you through the UI steps or configuration settings!
